com.unity.render-pipelines.universal/Runtime/Decal/Entities/DecalEntityManager.cs

```diff
@@ -0,0 +1,446 @@
+        public List<DecalCulledChunk> culledChunks = new List<DecalCulledChunk>();
+        public List<DecalDrawCallChunk> drawCallChunks = new List<DecalDrawCallChunk>();
+        public int chunkCount;
+
+        private ProfilingSampler m_AddDecalSampler;
+        private ProfilingSampler m_ResizeChunks;
+        private ProfilingSampler m_SortChunks;
+
+        private DecalEntityIndexer m_DecalEntityIndexer = new DecalEntityIndexer();
+        private Dictionary<Material, int> m_MaterialToChunkIndex = new Dictionary<Material, int>();
+
+        private struct CombinedChunks
+        {
+            public DecalEntityChunk entityChunk;
+            public DecalCachedChunk cachedChunk;
+            public DecalCulledChunk culledChunk;
+            public DecalDrawCallChunk drawCallChunk;
+            public int previousChunkIndex;
+        }
+        private List<CombinedChunks> m_CombinedChunks = new List<CombinedChunks>();
+        private List<int> m_CombinedChunkRemmap = new List<int>();
+
+        private Material m_ErrorMaterial;
+        public Material errorMaterial => m_ErrorMaterial;
+
+        public DecalEntityManager()
+        {
+            m_AddDecalSampler = new ProfilingSampler("DecalEntityManager.CreateDecalEntity");
+            m_ResizeChunks = new ProfilingSampler("DecalEntityManager.ResizeChunks");
+            m_SortChunks = new ProfilingSampler("DecalEntityManager.SortChunks");
+            m_ErrorMaterial = new Material(Shader.Find("Hidden/InternalErrorShader"));
+        }
+
+        public bool IsValid(DecalEntity decalEntity)
+        {
+            return m_DecalEntityIndexer.IsValid(decalEntity);
+        }
+
+        public DecalEntity CreateDecalEntity(DecalProjector decalProjector)
+        {
+            var material = decalProjector.material;
+            if (material == null)
+                material = errorMaterial;
+
+            using (new ProfilingScope(null, m_AddDecalSampler))
+            {
+                int chunkIndex = CreateChunkIndex(material);
+                int entityIndex = entityChunks[chunkIndex].count;
+
+                DecalEntity entity = m_DecalEntityIndexer.CreateDecalEntity(entityIndex, chunkIndex);
+
+                DecalEntityChunk entityChunk = entityChunks[chunkIndex];
+                DecalCachedChunk cachedChunk = cachedChunks[chunkIndex];
+                DecalCulledChunk culledChunk = culledChunks[chunkIndex];
+                DecalDrawCallChunk drawCallChunk = drawCallChunks[chunkIndex];
+
+                // Make sure we have space to add new entity
+                if (entityChunks[chunkIndex].capacity == entityChunks[chunkIndex].count)
+                {
+                    using (new ProfilingScope(null, m_ResizeChunks))
+                    {
+                        int newCapacity = entityChunks[chunkIndex].capacity + entityChunks[chunkIndex].capacity;
+                        newCapacity = math.max(250, newCapacity);
```

Comment:
It is max, I am going to reduce min size to 128. Idea is that it starts with 128 and grows 2x. I think it is safer to do 2x scale than const, my reasons:
- Contant is super slow in huge decal numbers.
- I dont expect that many grows so it should not be issue getting out of hand.
- NativeArrays resized with unitialized memorty, it means even if we have big array it is not mapped on physical memory.

Reply:
> NativeArrays resized with unitialized memorty, it means even if we have big array it is not mapped on physical memory.
I am not sure I follow. I might be missing some knowledge here. Here is my understanding; when allocating the array we reserve a chunk of memory corresponding to the length of the array. Regardless of what elements in the array we use.  Also no one else can use this chunk of memory until we free it again by disposing the NativeArray. So we still end up reserving physical memory?

We can allocate native arrays using [UninitializedMemory](https://docs.unity3d.com/ScriptReference/Unity.Collections.NativeArrayOptions.html) which skips the initialization of the memory to zero. Currently, the decal system does clear the memory as it is using the default option.

Reply:
Also as each material has its own chunks this ends up taking a sizeable amount of memory;
`decal materials (each material has its own chunk) * 128 (min size) * 492 bytes (per entity data stored across the 4 chunk types).`

Reply:
Thats good point, going to change the allocation with uninit

Reply:
Even with uninitialized memory the device will still reserve the memory to my knowledge.

Also be very careful with uninit. I think in your case it could be ok as you keep count externally. But if any section checks `IsValid()` on a part of the array you havent written to before it could return `true` by sheer coincidence.

Reply:
> NativeArrays resized with unitialized memorty, it means even if we have big array it is not mapped on physical memory.
Many mobile devices only have 4 GB of memory per process due to 32 bit virtual memory spaces, so it's best not to be wasteful. iOS included: https://developer.apple.com/library/archive/documentation/Performance/Conceptual/ManagingMemory/Articles/AboutMemory.html

Reply:
Changed to uninit allocation